### PR TITLE
[DO NOT MERGE, USE #29] rates fetcher interface

### PR DESCRIPTION
**Pull request by tiero**
This commits adds an idea of what could look like an abstract interface where the UI components (ie. ChooseTradingPair) could depend upon it.

Each Swap provider will implement in the most sensible and specific way, in the end the UI will need to convert from Amount A to Amount B, based on what the user is entering.

The idea is to have one main method, which I called here `Preview`, that will calculate the "opposite" amount using the rates of the provider to make the conversion and the final calculation to pre-fill the other input. The second parameter is the boolean `isSend` which could be helpful for implementing the "what I need to send to receive X of this coin"
I added also `PreviewGivenSend` and `PreviewGivenReceive`, which are single parameter version of `Preview`. They are redundant, but just left there as alternative in case the isSend boolean in Preview is not enough self-documenting.

The reason I didn't exposed a simple method that returns the rates/price is that it will uplift underlying implementation details of specific swap providers in the UI, which should be abstracted away if possible.

BONUS: I added a `ExampleFetcher` that is a dumb CoinGecko wrapper that shows how it could be implemented. IMO should not be merged, just there as showcase.

## TODOs

- [x] Refactor ChooseTradingPair to trigger the Preview method `onAmountChange` and `onAssetChange` binding specific implementation to currency pairs

Please @opendexnetwork/opendex-app review
﻿

### CODE REVIEW

**Comment by tiero** [comment]
@opendexnetwork/opendex-app I removed `Preview` method as agreed. I left there examples `ratesExample.ts` and `ratesExampleHook.ts` maybe could be beneficial for future providers as reference.

File: package.json [hunk 1]
```diff
@@ -16,6 +16,7 @@
     "bignumber.js": "^9.0.0",
     "bitcoinjs-lib": "^5.1.10",
     "boltz-core": "^0.4.0",
+    "decimal.js": "^10.2.1",
```
Comment (maybeast) [inline package.json:19]: Any particular reason we need `decimal.js` as we already have `bignumber.js` as a dependency?

Comment (tiero) [reply package.json:19]: Ok fine by me, as long we can handling floating point numbers we can move to bignumber.js

Comment (maybeast) [reply package.json:19]: Yup, we can. It's pretty much the same thing apart from slightly different API.

Comment (tiero) [reply package.json:19]: Done.

Comment (thomaseizinger) [reply package.json:19]: For reference: https://github.com/MikeMcl/big.js/issues/45

**Comment by thomaseizinger** [comment]
> @opendexnetwork/opendex-app I removed `Preview` method as agreed. I left there examples `ratesExample.ts` and `ratesExampleHook.ts` maybe could be beneficial for future providers as reference.

@tiero I think you forgot to push :)

**Comment by tiero** [comment]
@thomaseizinger Moving the repo from private to public, made my fork getting "disconnected" somehow from the PR. Let me sync again

**Comment by thomaseizinger** [comment]
Closing in favor of #29.

**Pull request closed without merging**